src/analysis/processing/qgsalgorithmexportgeometryattributes.cpp

```diff
@@ -0,0 +1,273 @@
+  if ( !source )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  const int method = parameterAsEnum( parameters, QStringLiteral( "METHOD" ), context );
+
+  const Qgis::WkbType wkbType = source->wkbType();
+  QgsFields fields = source->fields();
+  QgsFields newFields;
+
+  bool exportZ = false;
+  bool exportM = false;
+  if ( QgsWkbTypes::geometryType( wkbType ) == Qgis::GeometryType::Polygon )
+  {
+    newFields.append( QgsField( QStringLiteral( "area" ), QMetaType::Type::Double ) );
+    newFields.append( QgsField( QStringLiteral( "perimeter" ), QMetaType::Type::Double ) );
+  }
+  else if ( QgsWkbTypes::geometryType( wkbType ) == Qgis::GeometryType::Line )
+  {
+    newFields.append( QgsField( QStringLiteral( "length" ), QMetaType::Type::Double ) );
+    if ( !QgsWkbTypes::isMultiType( wkbType ) )
+    {
+      newFields.append( QgsField( QStringLiteral( "straightdis" ), QMetaType::Type::Double ) );
+      newFields.append( QgsField( QStringLiteral( "sinuosity" ), QMetaType::Type::Double ) );
+    }
+  }
+  else
+  {
+    if ( QgsWkbTypes::isMultiType( wkbType ) )
+    {
+      newFields.append( QgsField( QStringLiteral( "numparts" ), QMetaType::Type::Int ) );
+    }
+    else
+    {
+      newFields.append( QgsField( QStringLiteral( "xcoord" ), QMetaType::Type::Double ) );
+      newFields.append( QgsField( QStringLiteral( "ycoord" ), QMetaType::Type::Double ) );
+      if ( QgsWkbTypes::hasZ( wkbType ) )
+      {
+        newFields.append( QgsField( QStringLiteral( "zcoord" ), QMetaType::Type::Double ) );
+        exportZ = true;
+      }
+      if ( QgsWkbTypes::hasM( wkbType ) )
+      {
+        newFields.append( QgsField( QStringLiteral( "mvalue" ), QMetaType::Type::Double ) );
+        exportM = true;
+      }
+    }
+  }
+
+  fields = QgsProcessingUtils::combineFields( fields, newFields );
+
+  QString dest;
+  std::unique_ptr<QgsFeatureSink> sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields, wkbType, source->sourceCrs() ) );
+  if ( !sink )
+  {
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+  }
+
+  QgsCoordinateTransform transform;
+  mDa = QgsDistanceArea();
+
+  if ( method == 2 )
+  {
+    mDa.setSourceCrs( source->sourceCrs(), context.transformContext() );
+    mDa.setEllipsoid( context.ellipsoid() );
+    mDistanceConversionFactor = QgsUnitTypes::fromUnitToUnitFactor( mDa.lengthUnits(), context.distanceUnit() );
+    mAreaConversionFactor = QgsUnitTypes::fromUnitToUnitFactor( mDa.areaUnits(), context.areaUnit() );
+  }
+  else if ( method == 1 )
+  {
+    if ( !context.project() )
+    {
+      throw QgsProcessingException( QObject::tr( "No project is available in this context" ) );
+    }
+    transform = QgsCoordinateTransform( source->sourceCrs(), context.project()->crs(), context.project() );
```

Comment:
For thread safety we should retrieve the project crs in the prepareAlgorithm stage